x11: Handle external fullscreen toggles

eduardosm opened the pull request:
Some window managers let you change the fullscreen state of any window using a key combination. For example, on XFWM you can use Alt+F11 and on Compiz you can configure a key combination with the "Extra WM actions" plugin.
With this change mpv will handle these fullscreen state changes. So, if you enter into fullscreen mode using the WM's shortcut and then you use mpv's fullscreen toggle, you will get back into window mode.

wm4 commented:
I made some comments. In general it looks ok. I'd prefer if this code was in a separate function, instead of inside vo_x11_check_events itself.

eduardosm commented:
I removed the event I had added and now it uses MP_EVENT_WIN_STATE. I don't use vo_x11_fullscreen because I think will save the fullscreen dimensions as windowed dimensions (the code after "// Save old state before entering fullscreen").

eduardosm commented:
The code I added would just restore the window dimensions when exiting fullscreen mode and update x11->fs and opts->fullscreen.

wm4 commented:
I think you also need to reset x11->pos_changed_during_fs and size_changed_during_fs when leaving fullscreen, not sure. Now that vo_x11_check_net_wm_state_fullscreen_change us a separate function, you probably can move it somewhere so that no vo_x11_move_resize forward-declaration is needed.

eduardosm commented:
vo_x11_move_resize is defined after vo_x11_check_events. So, unless we move vo_x11_move_resize to somewhere before vo_x11_check_events, we will need a forward declaration for vo_x11_move_resize or vo_x11_check_net_wm_state_fullscreen_change.

wm4 commented:
OK, then never mind.

wm4 commented:
Actually, still move the forward declaration to the beginning of the file if you haven't already.

eduardosm commented:
Done.

wm4 commented:
Finally, do you agree that your contribution might be relicensed to LGPL v2.1 or later? (Which may or may not happen - for now changing to LGPL is just an idea.)

eduardosm commented:
Yes, that's not problem for me.

wm4 commented:
Squashed, fixed the commit message (72 line limit), and pushed.
